File: interpreter/go/internal/stack/stack.go
package stack

import "mop-lang/interpreter/go/internal/errors"

// Stack represents a stack data structure with overflow and underflow checking
type Stack struct {
	buffer       []float64
	stackPointer int
	size         int
}

// New creates a new stack with the specified size
func New(size int) *Stack {
	return &Stack{
		buffer:       make([]float64, size),
		stackPointer: -1,
		size:         size,
	}
}

// Push adds a number to the top of the stack
func (s *Stack) Push(number float64) {
	// Check for stack overflow
	if s.stackPointer+1 >= s.size {
		errors.HandleStackOverflow()
	}
	s.stackPointer++
	s.buffer[s.stackPointer] = number
}

// Pop removes and returns the top element from the stack
func (s *Stack) Pop() float64 {
	// Check for stack underflow
	if s.stackPointer < 0 {
		errors.HandleStackUnderflow("cannot pop from empty stack")
	}
	number := s.buffer[s.stackPointer]
	s.stackPointer--
	return number
}

// Top returns the top element of the stack without removing it
func (s *Stack) Top() float64 {
	// Check for stack underflow
	if s.stackPointer < 0 {
		errors.HandleStackUnderflow("cannot access top of empty stack")
	}
	return s.buffer[s.stackPointer]
}

// SizeCheck verifies that the stack has enough elements
func (s *Stack) SizeCheck(requiredElements int) {
	// Check if stack has enough elements
	if s.stackPointer+1 < requiredElements {
		errors.HandleStackUnderflowWithCount(requiredElements, s.stackPointer+1)
	}
}

// CurrentSize returns the current number of elements in the stack
func (s *Stack) CurrentSize() int {
	return s.stackPointer + 1
}

// Size returns the maximum capacity of the stack
func (s *Stack) Size() int {
	return s.size
}

File: interpreter/python/mopl/stack.py
"""Stack implementation for the MOPLang interpreter."""

from .error_handling import handle_stack_overflow, handle_stack_underflow, handle_stack_underflow_with_count


class Stack:
    """Stack implementation with overflow and underflow checking."""
    
    def __init__(self, size):
        self.buffer = [0 for _ in range(size)]  # create buffer for the specified stack size and fill it with 0
        self.stack_pointer = -1  # start sp at -1 so that we can properly start with 0, as it always should be
        self.size = size

    def push(self, number):
        """Push a number onto the stack."""
        # Check for stack overflow
        if self.stack_pointer + 1 >= self.size:
            handle_stack_overflow()
        self.stack_pointer += 1
        self.buffer[self.stack_pointer] = number

    def pop(self):
        """Pop a number from the stack."""
        # Check for stack underflow
        if self.stack_pointer < 0:
            handle_stack_underflow("pop from empty stack")
        number = self.buffer[self.stack_pointer]
        self.stack_pointer -= 1
        return number
    
    def top(self):
        """Get the top element of the stack without removing it."""
        # Check for stack underflow
        if self.stack_pointer < 0:
            handle_stack_underflow("access top of empty stack")
        return self.buffer[self.stack_pointer]
    
    def size_check(self, required_elements):
        """Check if stack has enough elements."""
        if self.stack_pointer + 1 < required_elements:
            handle_stack_underflow_with_count(required_elements, self.stack_pointer + 1)

    def current_size(self):
        """Get the current number of elements in the stack."""
        return self.stack_pointer + 1
